### PR TITLE
Support for custom providers

### PR DESCRIPTION
**Pull request by walkermatt**
As discussed in #123 this PR adds support for custom providers. I've included an example (`/examples/custom-provider.js`) which defines a custom provider for OS OpenNames (using https://t0.ads.astuntechnology.com/open/search/osopennames/)

Example via RawGit: https://rawgit.com/walkermatt/ol-geocoder/custom-providers/examples/custom-provider.html

I'm passing the custom provider instance via the `provider` property of the Geocoder constructor options parameter which feels fairly clean. This means that there are no external API changes apart from the value of the `provider` property being either `{string|object}` where the `object` exposes the methods `getParameters` and `handleResponse`.

### CODE REVIEW

File: src/js/nominatim.js [hunk 1]
```diff
@@ -111,7 +115,7 @@ export class Nominatim {
     ajax.url = document.location.protocol + provider.url;
     ajax.data = provider.params;
 
-    if (options.provider === C.providers.BING) {
+    if (provider.callbackName) {
```
Comment (walkermatt) [inline src/js/nominatim.js]: Enable JSONP if the provider defines a `callbackName` in the object returned by `getParameters`.

**Comment by jonataswalker** [comment]
Hey @walkermatt would you like to be a contributor in this project? I'm not having enough time to dedicate here.

**Comment by walkermatt** [comment]
I'd be happy to contribute although I also find a challenge to find the time between work and family :smile:.

**Comment by jonataswalker** [comment]
Great, I've sent you an invitation.

**Comment by walkermatt** [comment]
Thanks @jonataswalker, are you happy with me merging this PR?

**Comment by jonataswalker** [comment]
Yeah, @walkermatt please go ahead. Thanks for your help.

**Comment by moccand** [comment]
Hi,

I checked with French National Addresses WebServices (ex : http://api-adresse.data.gouv.fr/search?q=224%20rue%20de&lat=48.86&lon=2.35) and it works nicely.

Thanks !

Guillaume

**Comment by moccand** [comment]
here is my provider : 

```
/**
   * Custom provider for French Base Adresse Nationnale (BAN) covering France
   * Factory function which returns an object with the methods getParameters
   * and handleResponse called by the Geocoder
   */
  function BanFrSearch (options) {
    var url = options.url;
    var lat = options.lat;
    var lon = options.lon;
    return {
      /**
       * Get the url, query string parameters and optional JSONP callback
       * name to be used to perform a search.
       * @param {object} options Options object with query, key, lang,
       * countrycodes and limit properties.
       * @return {object} Parameters for search request
       */
      getParameters: function (options) {
        return {
          url: url,
          //callbackName: 'callback',
          params: {
            q: options.query,
            lat:lat,
            lon: lon
          }
        };
      },
      /**
       * Given the results of performing a search return an array of results
       * @param {object} data returned following a search request
       * @return {Array} Array of search results
       */
      handleResponse: function (results) {

        // The API returns a GeoJSON FeatureCollection
        if (results && results.features && results.features.length) {
            return results.features.map(function (feature) {
            return {
              lon: feature.geometry.coordinates[0],
              lat: feature.geometry.coordinates[1],
              address: {
                // Simply return a name in this case, could also return road,
                // building, house_number, city, town, village, state,
                // country
                name: feature.properties.name,
                postcode: feature.properties.postcode,
                city: feature.properties.city,
                //state: feature.properties.context,
                //country: "France"
              }
            };
          });
        } else {
          return;
        }
      }
    };
  }    
    
    
  // Create an instance of the custom provider, passing any options that are
  // required
  var provider = BanFrSearch({
    url: '//api-adresse.data.gouv.fr/search',
    lat: 48.86,
    lon: 2.35
  });   

```

**Comment by filipheymans** [comment]
Tested with custom provider; works perfect!

**Pull request merged**